test/Gate_test.jl

```diff
@@ -342,6 +409,25 @@
         end
     end
 
+    @testset "random unitary" begin
+        # test_throws on a non-unitary matrix
+        @test_throws ArgumentError SU{4}(1, 2; array = rand(ComplexF32, 4, 4), N = 2)
+
+        # test_throws on a non-square matrix
+        @test_throws ArgumentError SU{4}(1, 2; array = rand(ComplexF32, 4, 2), N = 2)
+
+        # test_throws on a matrix without size (N, N)
+        @test_throws ArgumentError SU{4}(1, 2; array = rand(ComplexF32, 2, 2), N = 2)
+
+        # test_throws SU{N} with N not a power of 2
+        @test_throws ArgumentError SU{3}(1, 2; array = rand(ComplexF32, 3, 3), N = 2)
+
+        # test_throws when there are not log2(N) lanes
+        rand_matrix = rand(ComplexF32, 4, 4)
+        q, _ = qr(rand_matrix)
+        @test_throws ArgumentError SU{4}(1, 2, 3; array = Matrix{ComplexF32}(q), N = 2)
```

Comment:
These tests are not correct with `N`.

```suggestion
        @test_throws ArgumentError SU{4}(1, 2; array = rand(ComplexF32, 4, 4), N = 2)

        # test_throws on a non-square matrix
        @test_throws ArgumentError SU{4}(1, 2; array = rand(ComplexF32, 4, 2), N = 2)

        # test_throws on a matrix without size (N, N)
        @test_throws ArgumentError SU{4}(1, 2; array = rand(ComplexF32, 2, 2), N = 2)

        # test_throws SU{N} with N not a power of 2
        @test_throws ArgumentError SU{3}(1, 2; array = rand(ComplexF32, 3, 3), N = 2)

        # test_throws when there are not log2(N) lanes
        rand_matrix = rand(ComplexF32, 4, 4)
        q, _ = qr(rand_matrix)
        @test_throws ArgumentError SU{4}(1, 2, 3; array = Matrix{ComplexF32}(q), N = 2)
```